sparse/unit_test/Test_Sparse_extractCrsDiagonalBlocks.hpp

```diff
@@ -0,0 +1,156 @@
+
+  crsMat_t A;
+  std::vector<crsMat_t> DiagBlks(nblocks);
+
+  if (nrows != 0) {
+    // Generate test matrix
+    const size_type nnz = 2 + (nrows - 2) * 3 + 2;
+    RowMapType_hm hrow_map("hrow_map", nrows + 1);
+    EntriesType_hm hentries("hentries", nnz);
+    ValuesType_hm hvalues("hvalues", nnz);
+
+    // first row
+    hrow_map(0) = 0;
+    hentries(0) = 0;
+    hentries(1) = 1;
+    hvalues(0)  = 0;
+    hvalues(1)  = 1;
+    // rows in between
+    int cnt = 2;
+    for (int i = 1; i <= (nrows - 2); i++) {
+      hrow_map(i)       = cnt;
+      hentries(cnt)     = -1 + i;
+      hentries(cnt + 1) = 0 + i;
+      hentries(cnt + 2) = 1 + i;
+      hvalues(cnt)      = -1 + i;
+      hvalues(cnt + 1)  = 0 + i;
+      hvalues(cnt + 2)  = 1 + i;
+      cnt += 3;
+    }
+    // last row
+    hrow_map(nrows - 1) = cnt;
+    hentries(nnz - 2)   = nrows - 2;
+    hentries(nnz - 1)   = nrows - 1;
+    hvalues(nnz - 2)    = nrows - 2;
+    hvalues(nnz - 1)    = nrows - 1;
+    // last element of row_map
+    hrow_map(nrows) = nnz;
+
+    // Allocate A on device memory
+    RowMapType row_map("row_map", nrows + 1);
+    EntriesType entries("entries", nnz);
+    ValuesType values("values", nnz);
+
+    // Copy from host to device
+    Kokkos::deep_copy(row_map, hrow_map);
+    Kokkos::deep_copy(entries, hentries);
+    Kokkos::deep_copy(values, hvalues);
+
+    // Construct a CRS matrix
+    A = crsMat_t("CrsMatrix", nrows, nrows, nnz, values, row_map, entries);
+  }
+
+  // Extract
+  KokkosSparse::Impl::kk_extract_diagonal_blocks_crsmatrix_sequential(A,
+                                                                      DiagBlks);
+
+  // Checking
+  lno_t numRows = 0;
+  lno_t numCols = 0;
+  for (int i = 0; i < nblocks; i++) {
+    numRows += DiagBlks[i].numRows();
+    numCols += DiagBlks[i].numCols();
+  }
+
+  EXPECT_TRUE(numRows == static_cast<lno_t>(nrows));
+  EXPECT_TRUE(numCols == static_cast<lno_t>(nrows));
+
+  if (nrows > 0) {
+    bool flag       = true;
+    lno_t col_start = 0;
+    for (int i = 0; i < nblocks; i++) {
+      RowMapType_hm hrow_map_diagblk("hrow_map_diagblk",
+                                     DiagBlks[i].numRows() + 1);
+      EntriesType_hm hentries_diagblk("hentries_diagblk", DiagBlks[i].nnz());
+      ValuesType_hm hvalues_diagblk("hvalues_diagblk", DiagBlks[i].nnz());
+
+      Kokkos::deep_copy(hrow_map_diagblk, DiagBlks[i].graph.row_map);
+      Kokkos::deep_copy(hentries_diagblk, DiagBlks[i].graph.entries);
+      Kokkos::deep_copy(hvalues_diagblk, DiagBlks[i].values);
+
+      for (int j = 0; j < static_cast<int>(DiagBlks[i].numRows()); j++) {
+        size_type k1 = hrow_map_diagblk(j);
+        size_type k2 = hrow_map_diagblk(j + 1);
+        for (size_type k = k1; k < k2; k++) {
+          scalar_t col = static_cast<scalar_t>(hentries_diagblk(k) + col_start);
+          scalar_t val = hvalues_diagblk(k);
+          if (Kokkos::abs(col - val) != 0) {
```

Comment:
Shouldn't we compare `hvalues` to `DiagBlks[?].values` here?

Reply:
@e10harvey See my comment above.